src/execution/execute.js

```diff
@@ -1026,15 +1122,34 @@ function defaultResolveTypeFn(
   value: mixed,
   context: mixed,
   info: GraphQLResolveInfo,
-  abstractType: GraphQLAbstractType
-): ?GraphQLObjectType {
+  abstractType: GraphQLAbstractType,
+): ?GraphQLObjectType | ?Promise<?GraphQLObjectType> {
   const possibleTypes = info.schema.getPossibleTypes(abstractType);
+  const promisedIsTypeOfResults = [];
+  const promisedIsTypeOfResultTypes = [];
+
   for (let i = 0; i < possibleTypes.length; i++) {
     const type = possibleTypes[i];
-    if (type.isTypeOf && type.isTypeOf(value, context, info)) {
-      return type;
+
+    if (type.isTypeOf) {
+      const isTypeOfResult = type.isTypeOf(value, context, info);
+
+      if (isThenable(isTypeOfResult)) {
+        promisedIsTypeOfResults.push(isTypeOfResult);
+        promisedIsTypeOfResultTypes.push(type);
+      } else if (isTypeOfResult) {
+        return type;
+      }
     }
   }
+
+  if (promisedIsTypeOfResults.length) {
+    return Promise.all(promisedIsTypeOfResults).then(isTypeOfResults => (
+      promisedIsTypeOfResultTypes[
+        isTypeOfResults.findIndex(isTypeOf => isTypeOf)
+      ]
+    ));
+  }
```

Comment:
I have a quick question, if one of these functions resolved true, but another one returns a rejected promise then the this will completely fail. Is this the expected behaviour?

While I think resolving all in parallel is a good idea, I was just thinking if the typeOf methods involve individual  database lookups or something similar then those kinds of implementations would be restricted to using the resolveType option to control synchronicity.